common/http-api-client/src/dns.rs

```diff
@@ -0,0 +1,153 @@
+//! DNS resolver configuration for internal lookups.
+//!
+//! The resolver itself is the set combination of the google, cloudflare, and quad9 endpoints
+//! supporting DoH and DoT.
+//!
+//! This resolver implements a fallback mechanism where, should the DNS-over-TLS resolution fail, a
+//! followup resolution will be done using the hosts configured default (e.g. `/etc/resolve.conf` on
+//! linux).
+//!
+//! Requires the `dns-over-https-rustls`, `webpki-roots` feature for the
+//! `hickory-resolver` crate
+
+use crate::ClientBuilder;
+
+use std::fmt;
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use hickory_resolver::{
+    config::{LookupIpStrategy, NameServerConfigGroup, ResolverConfig, ResolverOpts},
+    error::ResolveError,
+    lookup_ip::LookupIpIntoIter,
+    TokioAsyncResolver,
+};
+use once_cell::sync::OnceCell;
+use reqwest::dns::{Addrs, Name, Resolve, Resolving};
+use tracing::warn;
+
+impl ClientBuilder {
+    pub fn dns_resolver<R: Resolve + 'static>(mut self, resolver: Arc<R>) -> Self {
+        self.reqwest_client_builder = self.reqwest_client_builder.dns_resolver(resolver);
+        self
+    }
+}
+
+struct SocketAddrs {
+    iter: LookupIpIntoIter,
+}
+
+#[derive(Debug)]
+struct HickoryDnsSystemConfError(ResolveError);
+
+/// Wrapper around an `AsyncResolver`, which implements the `Resolve` trait.
+#[derive(Debug, Default, Clone)]
+pub struct HickoryDnsResolver {
+    /// Since we might not have been called in the context of a
+    /// Tokio Runtime in initialization, so we must delay the actual
+    /// construction of the resolver.
+    state: Arc<OnceCell<TokioAsyncResolver>>,
+    fallback: Arc<OnceCell<TokioAsyncResolver>>,
+}
+
+impl Resolve for HickoryDnsResolver {
+    fn resolve(&self, name: Name) -> Resolving {
+        let resolver = self.state.clone();
+        let fallback = self.fallback.clone();
+        Box::pin(async move {
+            let resolver = resolver.get_or_try_init(new_resolver)?;
+
+            // try the primary DNS resolver that we set up (DoH or DoT or whatever)
+            let lookup = match resolver.lookup_ip(name.as_str()).await {
+                Ok(res) => res,
+                Err(e) => {
+                    // on failure use the fall back system configured DNS resolver
+                    warn!("primary DNS failed w/ error {e}: using system fallback");
+                    let resolver = fallback.get_or_try_init(new_resolver_system)?;
+                    resolver.lookup_ip(name.as_str()).await?
+                }
+            };
+
+            let addrs: Addrs = Box::new(SocketAddrs {
+                iter: lookup.into_iter(),
+            });
+            Ok(addrs)
+        })
+    }
+}
+
+impl Iterator for SocketAddrs {
+    type Item = SocketAddr;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|ip_addr| SocketAddr::new(ip_addr, 0))
```

Comment:
is it safe to use port value of 0 here?

Reply:
I am not sure what the implications are actually. This comes from the `reqwest` implementation, but I will double check that this is safe to do in our context. To start though, the data is coming from the responses to DNS requests, so there isn't really any port to give them here (which SocketAddr requires). I am guessing this is implemented so that you can call things like `tokio::net::TcpStream::connect(...)` on the result.

I will investigate further.

Reply:
Okay so returning an `IntoIter<SockAddr>` is just part of interface that the `reqwest`  uses for dns resolution. This somehow has to do with reqwest internally using `hyper` / `tower` I think. So as long as the `HickoryDnsResolver` is used in conjunction with `reqwest` the ports are managed independently, and the `SockAddr`s with unset ports conforms to expected behavior.

To gain some separation from this I changed the public function to return the same iterator that hickory itself returns which is an `IntoIterator<IpAddr>`. So you have to go out of your way now by using the `reqwest` dns interface to get an iterator of `SockAddr`s (which is the expected behavior for `reqwest`), so shooting youself in the foot with unset ports should be pretty hard.